### PR TITLE
bpo-45847: Port _ctypes partly to PY_STDLIB_MOD (GH-29747)

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue45847

### CODE REVIEW

**Comment by tiran** [comment]
@ned-deily @ronaldoussoren I need your help with the macOS part of the patch. ``setup.py`` does some macOS specific things that I don't understand.

**Comment by ronaldoussoren** [comment]
Which parts do you need help with? 

It should be possible to move the macOS logic to configure.ac, although I'm a bit hazy at the moment about why some of the logic is there and some of the logic might be broken (in particular the code that sets USING_APPLE_OS_LIBFFI seems dodgy when self.use_system_libffi is false). 

I can look into this at the end of the week. 

BTW. Feel free to ping me directly, I don't regularly look at GitHub messages and only noticed your question promptly because Mail.app claimed it was junk mail.

File: configure.ac [hunk 1]
```diff
@@ -3124,6 +3124,60 @@ else
...
+
+    AC_CACHE_CHECK([for ffi_prep_closure_loc in ffi.h], [ac_cv_ffi_prep_closure_loc], [
+      AC_EGREP_HEADER([ffi_prep_closure_loc], [ffi.h], [ac_cv_ffi_prep_closure_loc=yes], [ac_cv_ffi_prep_closure_loc=no])
+    ])
+    AS_VAR_IF([ac_cv_ffi_prep_closure_loc], [yes], [
+      AC_DEFINE([HAVE_FFI_PREP_CLOSURE_LOC], [1],
+      [Define to 1 if you have the ffi_prep_closure_loc function in <ffi.h> header file.])
+    ])
+
+    AC_CACHE_CHECK([for fffi_closure_alloc in ffi.h], [ac_cv_ffi_closure_alloc], [
```
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
    AC_CACHE_CHECK([for ffi_closure_alloc in ffi.h], [ac_cv_ffi_closure_alloc], [
```

File: configure.ac [hunk 2]
```diff
@@ -3124,6 +3124,60 @@ else
...
+    AC_CHECK_HEADER([ffi.h], [
+      WITH_SAVE_ENV([
+        AC_CHECK_LIB([ffi], [ffi_call], [have_libffi=yes], [have_libffi=no])
+      ])
+    ])
+  ])
+], [
+  dnl private ffi copy
+  have_libffi=yes
+  LIBFFI_CFLAGS="-I\$(srcdir)/Modules/_ctypes/darwin -DUSING_MALLOC_CLOSURE_DOT_C=1 -DMACOSX"
```
Comment (erlend-aasland) [inline configure.ac:3140]: Is it safe to assume `-DMACOSX` here?

**Comment by pablogsal** [comment]
@tiran After this PR, the gentoo builder is failing:

https://buildbot.python.org/all/#/builders/58/builds/1217

```
test_free_called (test.test_code.CoExtra) ... Fatal Python error: Illegal instruction
Current thread 0xb7bd4700 (most recent call first):
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/test_code.py", line 575 in test_free_called
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/case.py", line 547 in _callTestMethod
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/case.py", line 591 in run
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/case.py", line 646 in __call__
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 122 in run
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 84 in __call__
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 122 in run
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 84 in __call__
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 122 in run
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/suite.py", line 84 in __call__
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/unittest/runner.py", line 197 in run
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/support/__init__.py", line 1009 in _run_suite
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/support/__init__.py", line 1135 in run_unittest
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/runtest.py", line 273 in _test_module
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/runtest.py", line 309 in _runtest_inner2
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/runtest.py", line 352 in _runtest_inner
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/runtest.py", line 227 in _runtest
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/runtest.py", line 257 in runtest
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/main.py", line 337 in rerun_failed_tests
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/main.py", line 715 in _main
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/main.py", line 658 in main
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/libregrtest/main.py", line 736 in main
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/test/__main__.py", line 2 in <module>
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/runpy.py", line 86 in _run_code
  File "/buildbot/buildarea/cpython/3.x.ware-gentoo-x86.nondebug/build/Lib/runpy.py", line 196 in _run_module_as_main
```

**Pull request merged**